File: app/src/main/java/com/home/young/myPassword/adapter/PasswordGroupAdapter.java
package com.home.young.myPassword.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import com.home.young.myPassword.R;
import com.home.young.myPassword.model.PasswordGroup;

public class PasswordGroupAdapter extends BaseAdapter {

    //region field
    private List<PasswordGroup> mPasswordGroups = new ArrayList<>();
    private Context mContext;
    private String mCurrentGroupName;
    //endregion

    //region constructor
    public PasswordGroupAdapter(Context context) {
        super();
        this.mContext = context;
    }
    //endregion

    //region function

    //region override
    @Override
    public int getCount() {
        return mPasswordGroups.size();
    }

    @Override
    public PasswordGroup getItem(int position) {
        return mPasswordGroups.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        ViewHolder viewHolder;
        if (convertView == null) {
            viewHolder = new ViewHolder();
            convertView = LayoutInflater.from(mContext).inflate(R.layout.password_group_item, null);
            viewHolder.name = (TextView) convertView.findViewById(R.id.fragment_password_group_name);
            viewHolder.arrowView = convertView.findViewById(R.id.fragment_password_group_arrow);
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) convertView.getTag();
        }

        PasswordGroup passwordGroup = getItem(position);
        viewHolder.name.setText(passwordGroup.getGroupName());

        if (passwordGroup.getGroupName().equals(mCurrentGroupName)) {
            viewHolder.arrowView.setVisibility(View.VISIBLE);
        } else {
            viewHolder.arrowView.setVisibility(View.INVISIBLE);
        }

        return convertView;
    }
    //endregion

    //region public
    public String getCurrentGroupName() {
        return mCurrentGroupName;
    }

    public void setCurrentGroupName(String currentGroupName) {
        this.mCurrentGroupName = currentGroupName;
        notifyDataSetChanged();
    }

    public void setData(List<PasswordGroup> passwordGroups) {
        this.mPasswordGroups.clear();
        this.mPasswordGroups.addAll(passwordGroups);
        notifyDataSetChanged();
    }

    public void addPasswordGroup(PasswordGroup passwordGroup) {
        mPasswordGroups.add(passwordGroup);
        notifyDataSetChanged();
    }

    public boolean removePasswordGroup(String passwordGroupName) {
        boolean result = false;
        for (int i = 0; i < mPasswordGroups.size(); i++) {
            PasswordGroup passwordGroup = mPasswordGroups.get(i);
            if (passwordGroup.getGroupName().equals(passwordGroupName)) {
                result = true;
                mPasswordGroups.remove(i);
                break;
            }
        }
        notifyDataSetChanged();
        return result;
    }
    //endregion

    //endregion

    //region nested class
    private class ViewHolder {
        TextView name;
        View arrowView;
    }
    //endregion
}

File: app/src/main/kotlin/org/tianjyan/pwd/adapter/PasswordGroupAdapter.kt
package org.tianjyan.pwd.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.TextView
import com.home.young.myPassword.R
import com.home.young.myPassword.model.PasswordGroup
import java.util.ArrayList

class PasswordGroupAdapter(private val mContext: Context) : BaseAdapter() {
    private val mPasswordGroups = ArrayList<PasswordGroup>()

    var currentGroupName: String? = null
        set(currentGroupName) {
            field = currentGroupName
            notifyDataSetChanged()
        }

    override fun getCount(): Int {
        return mPasswordGroups.size
    }

    override fun getItem(position: Int): PasswordGroup {
        return mPasswordGroups[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var convertView = convertView
        val viewHolder: ViewHolder
        if (convertView == null) {
            viewHolder = ViewHolder()
            convertView = LayoutInflater.from(mContext).inflate(R.layout.password_group_item, null)
            viewHolder.name = convertView!!.findViewById<View>(R.id.fragment_password_group_name) as TextView
            viewHolder.arrowView = convertView.findViewById(R.id.fragment_password_group_arrow)
            convertView.tag = viewHolder
        } else {
            viewHolder = convertView.tag as ViewHolder
        }

        val passwordGroup = getItem(position)
        viewHolder.name!!.text = passwordGroup.groupName

        if (passwordGroup.groupName == currentGroupName) {
            viewHolder.arrowView!!.visibility = View.VISIBLE
        } else {
            viewHolder.arrowView!!.visibility = View.INVISIBLE
        }

        return convertView
    }

    fun setData(passwordGroups: List<PasswordGroup>) {
        this.mPasswordGroups.clear()
        this.mPasswordGroups.addAll(passwordGroups)
        notifyDataSetChanged()
    }

    fun addPasswordGroup(passwordGroup: PasswordGroup) {
        mPasswordGroups.add(passwordGroup)
        notifyDataSetChanged()
    }

    fun removePasswordGroup(passwordGroupName: String): Boolean {
        var result = false
        for (i in mPasswordGroups.indices) {
            val passwordGroup = mPasswordGroups[i]
            if (passwordGroup.groupName == passwordGroupName) {
                result = true
                mPasswordGroups.removeAt(i)
                break
            }
        }
        notifyDataSetChanged()
        return result
    }

    private inner class ViewHolder {
        internal var name: TextView? = null
        internal var arrowView: View? = null
    }
}